src/shared/clients/ecsClient.ts

```diff
@@ -5,70 +5,76 @@
+}
+
+const MAX_RESULTS_PER_RESPONSE = 100
 export class DefaultEcsClient {
     public constructor(public readonly regionCode: string) {}
 
-    public async *listClusters(): AsyncIterableIterator<string> {
+    public async listClusters(nextToken?: string): Promise<EcsResourceAndToken> {
         const sdkClient = await this.createSdkClient()
-        const request: ECS.ListClustersRequest = {}
-        do {
-            const response = await this.invokeListClusters(request, sdkClient)
-            if (response.clusterArns) {
-                yield* response.clusterArns
+        const clusterArnList = await sdkClient
+            .listClusters({ maxResults: MAX_RESULTS_PER_RESPONSE, nextToken })
+            .promise()
+        if (clusterArnList.clusterArns?.length === 0) {
+            return { resource: [] }
+        }
+        try {
+            const clusterResponse = await sdkClient.describeClusters({ clusters: clusterArnList.clusterArns }).promise()
+            const response: EcsResourceAndToken = {
+                resource: clusterResponse.clusters!,
+                nextToken: clusterArnList.nextToken,
             }
-            request.nextToken = response.nextToken
-        } while (request.nextToken)
+            return response
+        } catch (error) {
+            getLogger().error('ecs: Failed to list clusters: %s', error)
+            throw error
+        }
     }
 
-    public async *listServices(cluster: string): AsyncIterableIterator<string> {
+    public async listServices(cluster: string, nextToken?: string): Promise<EcsResourceAndToken> {
         const sdkClient = await this.createSdkClient()
-        const request: ECS.ListServicesRequest = {
-            cluster,
+        const serviceArnList = await sdkClient
+            .listServices({ cluster: cluster, maxResults: MAX_RESULTS_PER_RESPONSE, nextToken })
+            .promise()
+        if (serviceArnList.serviceArns?.length === 0) {
+            return { resource: [] }
         }
-        do {
-            const response = await this.invokeListServices(request, sdkClient)
-            if (response.serviceArns) {
-                yield* response.serviceArns
+        try {
+            const serviceResponse = await sdkClient
+                .describeServices({ services: serviceArnList.serviceArns!, cluster: cluster })
+                .promise()
+            const response: EcsResourceAndToken = {
+                resource: serviceResponse.services!,
+                nextToken: serviceArnList.nextToken,
             }
-            request.nextToken = response.nextToken
-        } while (request.nextToken)
+            return response
+        } catch (error) {
+            getLogger().error('ecs: Failed to list services for cluster %s: %O', cluster, error)
+            throw error
+        }
     }
 
-    public async *listTaskDefinitionFamilies(): AsyncIterableIterator<string> {
+    public async listContainerNames(taskDefinition: string): Promise<string[]> {
```

Comment:
why are all of these functions being reverted from AsyncIterableIterator? Usually, AsyncIterableIterator is an improvement with no disadvantage. The caller can force AsyncIterableIterator to be a list .